agent/session.go

```diff
@@ -67,7 +67,14 @@ func newSession(ctx context.Context, agent *Agent, delay time.Duration, sessionI
 	)
 
 	if err != nil {
-		s.errs <- err
+		// since we are returning without launching the session goroutine, we
+		// need to provide the delay that is guaranteed by calling this
+		// function. We launch a goroutine so that we only delay the retry and
+		// avoid blocking the main loop.
+		go func() {
+			time.Sleep(delay)
+			s.errs <- err
```

Comment:
Just curious, what happens if the context gets done in the mean time ?

Reply:
Good point, I guess this wouldn't have a listener anymore.  Wondering if we should call `s.sendError(err)` instead, which selects between writing the error and the session closed channel being closed?

Reply:
Nothing: this is a buffered channel. If someone is listening, they pick it up. If no one is listening, it goes into the buffer immediately and the channel then gets garbage collected.